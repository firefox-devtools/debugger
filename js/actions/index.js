const breakpoints = require('./breakpoints');
const eventListeners = require('./event-listeners');
const sources = require('./sources');
const tabs = require('./tabs');

module.exports = Object.assign({}, breakpoints, eventListeners, sources, tabs);
