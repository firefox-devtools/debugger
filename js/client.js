const { DebuggerClient } = require('devtools/shared/client/main');
const { DebuggerTransport } = require('devtools/transport/transport');
const { TargetFactory } = require('devtools/client/framework/target');
const promise = require('devtools/sham/promise');

const socket = new WebSocket('ws://localhost:9000');
const transport = new DebuggerTransport(socket);
const client = new DebuggerClient(transport);

function connectToClient(onConnect) {
  client.connect().then(() => {
    return client.listTabs().then(onConnect);
  }).catch(err => console.log(err));
}

function connectToTab(tab, onNewSource) {
  let deferred = promise.defer();
  const options = { client, form: tab, chrome: false };

  TargetFactory.forRemoteTab(options).then(target => {
    target.activeTab.attachThread({}, (res, threadClient) => {
      threadClient.resume();
      window.gThreadClient = threadClient;
      deferred.resolve();
    });
  });

  return deferred.promise;
}

module.exports = {
  connectToClient,
  connectToTab
};
