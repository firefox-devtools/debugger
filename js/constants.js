/* -*- indent-tabs-mode: nil; js-indent-level: 2 -*- */
/* vim: set ft=javascript ts=2 et sw=2 tw=80: */
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
'use strict';

exports.UPDATE_EVENT_BREAKPOINTS = 'UPDATE_EVENT_BREAKPOINTS';
exports.FETCH_EVENT_LISTENERS = 'FETCH_EVENT_LISTENERS';

exports.TOGGLE_PRETTY_PRINT = 'TOGGLE_PRETTY_PRINT';
exports.BLACKBOX = 'BLACKBOX';

exports.ADD_BREAKPOINT = 'ADD_BREAKPOINT';
exports.REMOVE_BREAKPOINT = 'REMOVE_BREAKPOINT';
exports.ENABLE_BREAKPOINT = 'ENABLE_BREAKPOINT';
exports.DISABLE_BREAKPOINT = 'DISABLE_BREAKPOINT';
exports.SET_BREAKPOINT_CONDITION = 'SET_BREAKPOINT_CONDITION';

exports.ADD_SOURCE = 'ADD_SOURCE';
exports.LOAD_SOURCES = 'LOAD_SOURCES';
exports.LOAD_SOURCE_TEXT = 'LOAD_SOURCE_TEXT';
exports.SELECT_SOURCE = 'SELECT_SOURCE';
exports.UNLOAD = 'UNLOAD';
exports.RELOAD = 'RELOAD';

exports.ADD_TABS = 'ADD_TABS';
exports.SELECT_TAB = 'SELECT_TAB';
