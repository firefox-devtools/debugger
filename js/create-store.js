/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
'use strict';

const { createStore, applyMiddleware } = require('redux');
const { thunk } = require('devtools/client/shared/redux/middleware/thunk');
const { waitUntilService } = require('devtools/client/shared/redux/middleware/wait-service');
const { log } = require('devtools/client/shared/redux/middleware/log');
const { promise } = require('devtools/client/shared/redux/middleware/promise');
const { history } = require('devtools/client/shared/redux/middleware/history');

/**
 * This creates a dispatcher with all the standard middleware in place
 * that all code requires. It can also be optionally configured in
 * various ways, such as logging and recording.
 *
 * @param {object} opts:
 *        - log: log all dispatched actions to console
 *        - history: an array to store every action in. Should only be
 *                   used in tests.
 *        - middleware: array of middleware to be included in the redux store
 */
module.exports = (opts = {}) => {
  const middleware = [
    thunk,
    promise,

    // Order is important: services must go last as they always
    // operate on "already transformed" actions. Actions going through
    // them shouldn't have any special fields like promises, they
    // should just be normal JSON objects.
    waitUntilService
  ];

  if (opts.history) {
    middleware.push(history(opts.history));
  }

  if (opts.middleware) {
    opts.middleware.forEach(fn => middleware.push(fn));
  }

  if (opts.log) {
    middleware.push(log);
  }

  return applyMiddleware(...middleware)(createStore);
};
