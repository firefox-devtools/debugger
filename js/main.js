const React = require('react');
const ReactDOM = require('react-dom');
const { combineReducers, applyMiddleware, bindActionCreators } = require('redux');
const { Provider } = require('react-redux');
const configureStore = require('./create-store');
const reducers = require('./reducers');
const { connectToClient } = require('./client');
const actions = require('./actions');
const dom = React.DOM;
const TabList = React.createFactory(require('./components/TabList'));

const createStore = configureStore({ log: true });
const store = createStore(combineReducers(reducers));
window.store = store; // global for debugging purposes only!

connectToClient(response => {
  const tabs = response.tabs;
  store.dispatch(actions.newTabs(response.tabs));

  // if there's a pre-selected tab, connect to it and load the sources.
  // otherwise, just show the toolbox.
  if (hasSelectedTab()) {
    const selectedTab = getSelectedTab(tabs);
    store.dispatch(actions.selectTab({ tabActor: selectedTab.actor }))
      .then(() => store.dispatch(actions.loadSources()))
      .then(renderToolbox);
  } else {
    renderToolbox();
  }
});

/**
 * Check to see if the url hash has a selected tab
 * e.g. #tab=child2
 */
function hasSelectedTab() {
  return window.location.hash.includes('tab');
}

/**
 * get the selected tab from the url hash
 * e.g. #tab=child2
 *
 * tabs are keyed by their child id,
 * this is because the actor connection id increments every refresh and
 * tab id is always 1.
 *
 */
function getSelectedTab(tabs) {
  const childId = window.location.hash.split('=')[1];
  return tabs.find((tab) => tab.actor.includes(childId));
}

function renderToolbox() {
  ReactDOM.render(
    React.createElement(
      Provider,
      { store },
      TabList()
    ),
    document.querySelector('#mount')
  );
}
