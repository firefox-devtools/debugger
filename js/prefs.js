const { PrefsHelper } = require('devtools/client/shared/prefs');

/**
 * Shortcuts for accessing various debugger preferences.
 */
let Prefs = new PrefsHelper('devtools', {
  workersAndSourcesWidth: ['Int', 'debugger.ui.panes-workers-and-sources-width'],
  instrumentsWidth: ['Int', 'debugger.ui.panes-instruments-width'],
  panesVisibleOnStartup: ['Bool', 'debugger.ui.panes-visible-on-startup'],
  variablesSortingEnabled: ['Bool', 'debugger.ui.variables-sorting-enabled'],
  variablesOnlyEnumVisible: ['Bool', 'debugger.ui.variables-only-enum-visible'],
  variablesSearchboxVisible: ['Bool', 'debugger.ui.variables-searchbox-visible'],
  pauseOnExceptions: ['Bool', 'debugger.pause-on-exceptions'],
  ignoreCaughtExceptions: ['Bool', 'debugger.ignore-caught-exceptions'],
  sourceMapsEnabled: ['Bool', 'debugger.source-maps-enabled'],
  prettyPrintEnabled: ['Bool', 'debugger.pretty-print-enabled'],
  autoPrettyPrint: ['Bool', 'debugger.auto-pretty-print'],
  workersEnabled: ['Bool', 'debugger.workers'],
  editorTabSize: ['Int', 'editor.tabsize'],
  autoBlackBox: ['Bool', 'debugger.auto-black-box'],
  promiseDebuggerEnabled: ['Bool', 'debugger.promise']
});

module.exports = Prefs;
