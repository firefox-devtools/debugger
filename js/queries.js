
function getSource(state, actor) {
  return state.sources.sources[actor];
}

function getSources(state) {
  return state.sources.sources;
}

function getSourceCount(state) {
  return Object.keys(state.sources.sources).length;
}

function getSourceByURL(state, url) {
  for (let k in state.sources.sources) {
    const source = state.sources.sources[k];
    if (source.url === url) {
      return source;
    }
  }
}

function getSourceByActor(state, actor) {
  for (let k in state.sources.sources) {
    const source = state.sources.sources[k];
    if (source.actor === actor) {
      return source;
    }
  }
}

function getSelectedSource(state) {
  return state.sources.sources[state.sources.selectedSource];
}

function getSelectedSourceOpts(state) {
  return state.sources.selectedSourceOpts;
}

function getSourceText(state, actor) {
  return state.sources.sourcesText[actor];
}

function getBreakpoints(state) {
  return Object.keys(state.breakpoints.breakpoints).map(k => {
    return state.breakpoints.breakpoints[k];
  });
}

function getBreakpoint(state, location) {
  return state.breakpoints.breakpoints[makeLocationId(location)];
}

function getTabs(state) {
  return state.tabs.tabs;
}

function getSelectedTab(state) {
  return state.tabs.selectedTab;
}

function makeLocationId(location) {
  return location.actor + ':' + location.line.toString();
}

module.exports = {
  getSource,
  getSources,
  getSourceCount,
  getSourceByURL,
  getSourceByActor,
  getSelectedSource,
  getSelectedSourceOpts,
  getSourceText,
  getBreakpoint,
  getBreakpoints,
  getTabs,
  getSelectedTab,
  makeLocationId
};
