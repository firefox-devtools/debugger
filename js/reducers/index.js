/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
'use strict';

const eventListeners = require('./event-listeners');
const sources = require('./sources');
const breakpoints = require('./breakpoints');
const asyncRequests = require('./async-requests');
const tabs = require('./tabs');

module.exports = {
  eventListeners,
  sources,
  breakpoints,
  asyncRequests,
  tabs
};
