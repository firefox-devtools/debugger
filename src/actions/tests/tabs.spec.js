import {
  actions,
  selectors,
  createStore,
  makeSource
} from "../../utils/test-head";
const { getSelectedSource, getTabs, getSelectedTab } = selectors;

import { sourceThreadClient as threadClient } from "./helpers/threadClient.js";

describe("adding tabs", () => {});

describe("selecting tabs", () => {});

describe("closing tabs", () => {
  it("closing a tab", async () => {
    const { dispatch, getState } = createStore(threadClient);
    await dispatch(actions.newSource(makeSource("foo.js")));
    await dispatch(actions.selectLocation({ sourceId: "foo.js" }));

    await dispatch(actions.closeTab("foo.js"));

    expect(getSelectedSource(getState())).toBe(undefined);
    expect(getSelectedTab(getState())).toMatchSnapshot();
    expect(getTabs(getState()).size).toBe(0);
  });

  it("closing the inactive tab", async () => {
    const { dispatch, getState } = createStore(threadClient);
    await dispatch(actions.newSource(makeSource("foo.js")));
    await dispatch(actions.newSource(makeSource("bar.js")));
    await dispatch(actions.selectLocation({ sourceId: "foo.js" }));
    await dispatch(actions.selectLocation({ sourceId: "bar.js" }));

    await dispatch(actions.closeTab("foo.js"));

    expect(getSelectedSource(getState()).get("id")).toBe("bar.js");
    expect(getSelectedTab(getState())).toMatchSnapshot();
    expect(getTabs(getState()).size).toBe(1);
    dispatch(actions.closeTab("bar.js"));
  });

  it("closing the only tab", async () => {
    const { dispatch, getState } = createStore(threadClient);
    await dispatch(actions.newSource(makeSource("foo.js")));
    await dispatch(actions.selectLocation({ sourceId: "foo.js" }));

    await dispatch(actions.closeTab("foo.js"));

    expect(getSelectedSource(getState())).toBe(undefined);
    expect(getSelectedTab(getState())).toMatchSnapshot();
    expect(getTabs(getState()).size).toBe(0);
  });

  it("closing the active tab", async () => {
    const { dispatch, getState } = createStore(threadClient);
    await dispatch(actions.newSource(makeSource("foo.js")));
    await dispatch(actions.newSource(makeSource("bar.js")));
    await dispatch(actions.selectLocation({ sourceId: "foo.js" }));
    await dispatch(actions.selectLocation({ sourceId: "bar.js" }));

    await dispatch(actions.closeTab("bar.js"));

    expect(getSelectedSource(getState()).get("id")).toBe("foo.js");
    expect(getSelectedTab(getState())).toMatchSnapshot();
    expect(getTabs(getState()).size).toBe(1);
  });

  it("closing many inactive tabs", async () => {
    const { dispatch, getState } = createStore(threadClient);
    await dispatch(actions.newSource(makeSource("foo.js")));
    await dispatch(actions.newSource(makeSource("bar.js")));
    await dispatch(actions.newSource(makeSource("bazz.js")));
    await dispatch(actions.selectLocation({ sourceId: "foo.js" }));
    await dispatch(actions.selectLocation({ sourceId: "bar.js" }));
    await dispatch(actions.selectLocation({ sourceId: "bazz.js" }));

    await dispatch(actions.closeTabs(["foo.js", "bar.js"]));

    expect(getSelectedSource(getState()).get("id")).toBe("bazz.js");
    expect(getSelectedTab(getState())).toMatchSnapshot();
    expect(getTabs(getState()).size).toBe(1);
  });

  it("closing many tabs including the active tab", async () => {
    const { dispatch, getState } = createStore(threadClient);
    await dispatch(actions.newSource(makeSource("foo.js")));
    await dispatch(actions.newSource(makeSource("bar.js")));
    await dispatch(actions.newSource(makeSource("bazz.js")));
    await dispatch(actions.selectLocation({ sourceId: "foo.js" }));
    await dispatch(actions.selectLocation({ sourceId: "bar.js" }));
    await dispatch(actions.selectLocation({ sourceId: "bazz.js" }));

    await dispatch(actions.closeTabs(["bar.js", "bazz.js"]));

    expect(getSelectedSource(getState()).get("id")).toBe("foo.js");
    expect(getSelectedTab(getState())).toMatchSnapshot();
    expect(getTabs(getState()).size).toBe(1);
  });

  it("closing all the tabs", async () => {
    const { dispatch, getState } = createStore(threadClient);
    await dispatch(actions.newSource(makeSource("foo.js")));
    await dispatch(actions.newSource(makeSource("bar.js")));
    await dispatch(actions.selectLocation({ sourceId: "foo.js" }));
    await dispatch(actions.selectLocation({ sourceId: "bar.js" }));

    await dispatch(actions.closeTabs(["foo.js", "bar.js"]));

    expect(getSelectedSource(getState())).toBe(undefined);
    expect(getSelectedTab(getState())).toMatchSnapshot();
    expect(getTabs(getState()).size).toBe(0);
  });
});
