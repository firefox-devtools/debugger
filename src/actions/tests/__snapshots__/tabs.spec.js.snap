// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`closing tabs closing a tab 1`] = `
Object {
  "tab": undefined,
  "tabIndex": -1,
}
`;

exports[`closing tabs closing all the tabs 1`] = `
Object {
  "tab": undefined,
  "tabIndex": -1,
}
`;

exports[`closing tabs closing many inactive tabs 1`] = `
Object {
  "tab": Object {
    "id": "bazz.js",
    "title": "bazz.js",
    "tooltip": "http://localhost:8000/examples/bazz.js",
  },
  "tabIndex": 0,
}
`;

exports[`closing tabs closing many tabs including the active tab 1`] = `
Object {
  "tab": Object {
    "id": "foo.js",
    "title": "foo.js",
    "tooltip": "http://localhost:8000/examples/foo.js",
  },
  "tabIndex": 0,
}
`;

exports[`closing tabs closing the active tab 1`] = `
Object {
  "tab": Object {
    "id": "foo.js",
    "title": "foo.js",
    "tooltip": "http://localhost:8000/examples/foo.js",
  },
  "tabIndex": 0,
}
`;

exports[`closing tabs closing the inactive tab 1`] = `
Object {
  "tab": Object {
    "id": "bar.js",
    "title": "bar.js",
    "tooltip": "http://localhost:8000/examples/bar.js",
  },
  "tabIndex": 0,
}
`;

exports[`closing tabs closing the only tab 1`] = `
Object {
  "tab": undefined,
  "tabIndex": -1,
}
`;
